### PR TITLE
OTG Test - RT-5.2: Aggregate Interfaces

### LINKED ISSUE #558: OTG: RT-5.2: Aggregate Interfaces
- [README](https://github.com/openconfig/featureprofiles/blob/main/feature/interface/aggregate/otg_tests/aggregate_test/README.md)

### PR DESCRIPTION
**Pull request by octpetre**
The OTG conversion for **_RT-5.2: Aggregate Interfaces_**
The PR includes the changes from otgutils internal package raised here #514 
Test log output [here](https://drive.google.com/file/d/14Qjqr02oCzFfVX03Vz5v7_vkmqzGWqwF/view?usp=sharing)
Fixes #558

### CODE REVIEW

**Comment by dplore** [comment]
Resolves #558

**Comment by octpetre** [comment]
Fixes #558

File: feature/interface/aggregate/otg_tests/aggregate_test/aggregate_test.go [hunk 1]
```diff
@@ -353,6 +405,23 @@ func sortPorts(ports []*ondatra.Port) []*ondatra.Port {
 	return ports
 }
 
+// incrementedMac uses a mac string and increments it by the given i
+func incrementedMac(mac string, i int) (string, error) {
```
[the remark this answers is not in the document]
Comment (octpetre) [reply feature/interface/aggregate/otg_tests/aggregate_test/aggregate_test.go:409]: Corrected

File: feature/interface/aggregate/otg_tests/aggregate_test/aggregate_test.go [hunk 2]
```diff
@@ -244,42 +255,66 @@ func (tc *testCase) configureATE(t *testing.T) {
...
-	if p0.PMD() == ondatra.PMD100GFR {
-		i0.Ethernet().FEC().WithEnabled(false)
-	}
-	is100gfr := false
-	for _, p := range tc.atePorts[1:] {
-		if p.PMD() == ondatra.PMD100GFR {
-			is100gfr = true
+	tc.top.Ports().Add().SetName(p0.ID())
+	srcDev := tc.top.Devices().Add().SetName(ateSrc.Name)
+	srcEth := srcDev.Ethernets().Add().SetName(ateSrc.Name + ".Eth")
```
[the remark this answers is not in the document]
Comment (octpetre) [reply feature/interface/aggregate/otg_tests/aggregate_test/aggregate_test.go:260]: Yes, the names and mandatory in the model and they can't be duplicated

File: feature/interface/aggregate/otg_tests/aggregate_test/aggregate_test.go [hunk 3]
```diff
@@ -244,42 +255,66 @@ func (tc *testCase) configureATE(t *testing.T) {
...
+			if err != nil {
+				t.Fatal(err)
+			}
+			agg.Ports().Add().SetPortName(port.Name()).Ethernet().SetMac(newMac).SetName("LAGRx-" + strconv.Itoa(i))
+		}
+	} else {
+		agg.Protocol().SetChoice("lacp")
+		for i, p := range tc.atePorts[1:] {
+			port := tc.top.Ports().Add().SetName(p.ID())
+			newMac, err := incrementedMac(ateDst.MAC, i+1)
```
[the remark this answers is not in the document]
Comment (octpetre) [reply feature/interface/aggregate/otg_tests/aggregate_test/aggregate_test.go:286]: Yes I agree, I could add this to otgutils package which right now we use only for printing purposes. In that case I will open a separate admin PR to pull this function into main

File: feature/interface/aggregate/otg_tests/aggregate_test/aggregate_test.go [hunk 4]
```diff
@@ -244,42 +255,66 @@ func (tc *testCase) configureATE(t *testing.T) {
...
+			if err != nil {
+				t.Fatal(err)
+			}
+			lagPort := agg.Ports().Add().SetPortName(port.Name())
+			lagPort.Ethernet().SetMac(newMac).SetName("LAGRx-" + strconv.Itoa(i))
```
[the remark this answers is not in the document]
Comment (octpetre) [reply feature/interface/aggregate/otg_tests/aggregate_test/aggregate_test.go:291]: We had a discussion in the past to create high level gosnappi helper functions for something like this but it wasn't decided. All the tests are using this low level commands to create config. Ideally, the OTG should populate this names key by default.

File: feature/interface/aggregate/otg_tests/aggregate_test/aggregate_test.go [hunk 5]
```diff
@@ -244,42 +255,66 @@ func (tc *testCase) configureATE(t *testing.T) {
...
-	}
 
-	agg.IPv4().
-		WithAddress(ateDst.IPv4CIDR()).
-		WithDefaultGateway(dutDst.IPv4)
-	agg.IPv6().
-		WithAddress(ateDst.IPv6CIDR()).
-		WithDefaultGateway(dutDst.IPv6)
+	dstDev := tc.top.Devices().Add().SetName(agg.Name())
+	dstEth := dstDev.Ethernets().Add().SetName(ateDst.Name + ".Eth").SetPortName(agg.Name()).SetMac(ateDst.MAC)
```
[the remark this answers is not in the document]
Comment (octpetre) [reply feature/interface/aggregate/otg_tests/aggregate_test/aggregate_test.go:297]: Here "lag_name" is a connection choice option defined in the model. This is how we tie a device to a lag.

File: feature/interface/aggregate/otg_tests/aggregate_test/aggregate_test.go [hunk 6]
```diff
@@ -244,42 +255,66 @@ func (tc *testCase) configureATE(t *testing.T) {
...
+	ok := false
+	for n := 3; !ok; n-- {
+		if n == 0 {
+			t.Fatal("Not retrying ATE StartProtocols anymore.")
+		}
+		msg := testt.ExpectFatal(t, func(t testing.TB) {
+			t.Log("Trying ATE StartProtocols")
+			tc.ate.OTG().StartProtocols(t)
+			ok = true
+			t.Fatal("Success!")
```
[the remark this answers is not in the document]
Comment (octpetre) [reply feature/interface/aggregate/otg_tests/aggregate_test/aggregate_test.go:314]: I removed that entire block. I think this was added as a w/a earlier

**Review by marcushines — APPROVED** [review]

**Pull request merged**